add opentelemetry tracing instrumentation

sallyom opened the pull request.

frzifus commented on internal/tracing/tracing.go:
```diff
@@ -0,0 +1,181 @@
+	}, nil
+}
+
+// StartSpan creates a new span for routing proxy operations
+func StartSpan(ctx context.Context, operationName string, spanKind trace.SpanKind) (context.Context, trace.Span) {
+	tracer := otel.Tracer(ServiceName)
+	ctx, span := tracer.Start(ctx, operationName,
+		trace.WithSpanKind(spanKind),
+	)
+	return ctx, span
+}
+
+// StartHTTPSpan creates a new span for HTTP operations
+func StartHTTPSpan(ctx context.Context, operationName, method, route string) (context.Context, trace.Span) {
+	ctx, span := StartSpan(ctx, operationName, trace.SpanKindServer)
+	span.SetAttributes(
+		attribute.String(AttrHTTPMethod, method),
+		attribute.String(AttrHTTPRoute, route),
+	)
+	return ctx, span
+}
+
+// StartProxySpan creates a new span for proxy forwarding operations
+func StartProxySpan(ctx context.Context, operationName, connector string) (context.Context, trace.Span) {
+	ctx, span := StartSpan(ctx, operationName, trace.SpanKindClient)
+	span.SetAttributes(
+		attribute.String(AttrProxyConnector, connector),
+	)
+	return ctx, span
+}
```
Whats the reason for this wrapper? 

Seems the usage of it is also mixed (`interal/proxy/chat_completion.go`):
```golang
func (s *Server) chatCompletionsHandler(w http.ResponseWriter, r *http.Request) {
	ctx, span := tracing.StartHTTPSpan(r.Context(), tracing.OperationChatCompletions, r.Method, r.URL.Path)
	defer span.End()

	prefillPodURL := r.Header.Get(requestHeaderPrefillURL)
	requestID := r.Header.Get(requestHeaderRequestID)

	span.SetAttributes(
		attribute.String(tracing.AttrProxyConnector, s.connector),              <--
		attribute.String(tracing.AttrProxyDecoderURL, s.decoderURL.String()),   <--
	)
```

Maybe it makes sense to add some kind of http middleware?

frzifus commented on internal/tracing/tracing.go:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	if samplingRate := os.Getenv(envOTELSamplingRate); samplingRate != "" {
+		if rate, err := strconv.ParseFloat(samplingRate, 64); err == nil && rate >= 0.0 && rate <= 1.0 {
```
Does it make sense to inform the user e.g. by failing during the creation whenever a value is not recognized?

frzifus reviewed (commented):
Just left some comments :)

sallyom commented on internal/proxy/chat_completions.go:
```diff
@@ -29,13 +32,37 @@ var (
 	if prefillPodURL == "" {
 		s.logger.V(4).Info("skip disagreggated prefill")
+		span.SetAttributes(
+			attribute.Bool("llm_d.proxy.disaggregated_prefill", false),
```
yea was thinking that - this PR I gave the least love to

sallyom committed:
Fix connector logging bug: dereference pointer to show actual value

The connector variable is a *string from flag.String(), but the logger
was printing the pointer address instead of the actual string value.
This fixes the log output to show 'connector=nixlv2' instead of
'connector=0xc00012a1e0'.

sallyom committed:
Add minimal tracing instrumentation for auto-instrumentation

sallyom commented on internal/tracing/tracing.go:
```diff
@@ -0,0 +1,181 @@
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(config.ServiceName),
+			semconv.ServiceVersionKey.String("dev"),
```
updated to be more minimal tracing instrumentation, no longer relevant

sallyom committed:
add more spans

Signed-off-by: sallyom <somalley@redhat.com>

elevran commented:
@sallyom the routing-sidecar code has been moved to inference-scheduler repo (see this repo's README).
Can you please close your PR here and resubmit in the other repo if still needed?
